### PR TITLE
nrf52: Add USB peripheral driver

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds a USB peripheral driver for the nrf52840 MCU.

Makefile configuration is supplied for the nrf52840dk and the nrf52840-mdk. Feel free to suggest more boards if I missed some.

### Testing procedure

**WARNING**: due to errata 104 and errata 180, this driver won't work for the engineering sample A chips. From engineering sample B and newer it should be fine.

### Issues/PRs references

depends on #9830, requires #10916 for actual testing

image

### CODE REVIEW

**Comment by miri64** [comment]
Needs rebase, I believe ;-)

**Comment by bergzand** [comment]
> Needs rebase, I believe ;-)

Also needs a bit of work, I've neglected so far to rework this driver to the current periph_usbdev architecture.

**Comment by dylad** [comment]
BTW, I have a nrf52dk so I should be able to help here :)

**Comment by bergzand** [comment]
rebased and updated!

**Comment by bergzand** [comment]
Ready for review here

File: cpu/nrf52/include/nrfusb.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    size_t used;                            /**< Number of bytes from the
+                                                 buffer that are used           */
+    uint8_t buffer[NRF_USB_BUF_SPACE];      /**< Buffer space for endpoint data */
+    nrfusb_setup_state_t sstate;            /**< Setup request state machine    */
+} nrfusb_t;
+
+#ifdef __cplusplus
+}
+#endif
+#endif /* SAM_USB_H */
```
Comment (dylad) [inline cpu/nrf52/include/nrfusb.h]: Not SAM anymore

File: cpu/nrf52/periph/usbdev.c [hunk 2]
```diff
@@ -0,0 +1,654 @@
+/*
+ * Copyright (C) 2018 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sam0_common
```
Comment (dylad) [inline cpu/nrf52/periph/usbdev.c]: not sam0

File: cpu/nrf52/periph/usbdev.c [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+ * @{
+ * @file
+ * @brief   USB interface functions
+ *
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+#include <stdint.h>
+#include <stdlib.h>
+#include <errno.h>
```
Comment (dylad) [inline cpu/nrf52/periph/usbdev.c]: I didn't see any errno, maybe this can be removed ?

**Comment by bergzand** [comment]
>  Is there any restriction with nRF52832 devices ?? I see a bunch of reference to nRF52840 but does it work on nRF52832 ??

As far as I know, the nRF52832 doesn't have the USB peripheral, only the nRF52840 has one.

File: cpu/nrf52/periph/usbdev.c (hunk 3 above)
Comment (bergzand) [reply cpu/nrf52/periph/usbdev.c]: Required for the `-ENOTSUP` return values

File: cpu/nrf52/include/nrfusb.h (hunk 1 above)
Comment (bergzand) [reply cpu/nrf52/include/nrfusb.h]: Oh, :cry: 
fixed

File: cpu/nrf52/periph/usbdev.c (hunk 2 above)
Comment (bergzand) [reply cpu/nrf52/periph/usbdev.c]: Fixed

**Comment by bergzand** [comment]
@aabadie Would you mind taking a look here, @dylad doesn't have an nrf52840 based board available for testing?

**Comment by aabadie** [comment]
> the nRF52832 doesn't have the USB peripheral, only the nRF52840 has one

I checked the datasheet of both and this is indeed the case. nrf51 also doesn't provide an USB peripheral.

**Comment by aabadie** [comment]
I can test on nrf52840-dk/mdk.

**Comment by bergzand** [comment]
Squashed and rebased.

@aabadie I took the liberty of fixing the alphabetical order of the `Makefile.features` for the nrf52840-mdk board while fixing the merge conflict.

**Comment by bergzand** [comment]
And fixed two typo's in two commit messages

**Comment by aabadie** [comment]
> I took the liberty of fixing the alphabetical order of the Makefile.features for the nrf52840-mdk board while fixing the merge conflict.

Thanks @bergzand, I overlooked this when reviewing #11513.

Now that everything is green, let's merge this one. Awesome work :)

**Pull request merged**